ruby/ext/google/protobuf_c/upb.c

```diff
@@ -10134,7 +10138,7 @@ const unsigned short int __mon_yday[2][13] = {
 int64_t epoch(int year, int yday, int hour, int min, int sec) {
   int64_t years = year - EPOCH_YEAR;
 
-  int64_t leap_days = years / 4 - years / 100 + years / 400;
+  int64_t leap_days = isleap_count(year - 1) - isleap_count(EPOCH_YEAR - 1);
```

Comment:
Why not putting year - 1 into leap_count?

Reply:
Could you also add a test in https://github.com/protocolbuffers/protobuf/blob/7114d73dea3daeb88b378f53380b46a94beb05ae/conformance/binary_json_conformance_suite.cc#L2704

Reply:
Could you also port your code to https://github.com/protocolbuffers/upb/blob/d3762e96cb8910b8281419faee7bd70b323c753d/upb/json/parser.rl#L1707
and https://github.com/protocolbuffers/protobuf/blob/master/php/ext/google/protobuf/upb.c#L10137

Reply:
> Why not putting year - 1 into leap_count?

`static int leap_count(int year)` returns the number of leap days between 1 and year.
Therefore, this logic calculated the number of leap days between `1970` and `year - 1`. 

The relation between `is_leap` and `leap_count` is shown below. 

```
is_leap(1) => false
is_leap(2) => false
is_leap(3) => false
is_leap(4) => true
is_leap(5) => false
is_leap(6) => false
is_leap(7) => false
is_leap(8) => true

leap_count(1) => 0
leap_count(2) => 0
leap_count(3) => 0
leap_count(4) => 1
leap_count(5) => 1
leap_count(6) => 1
leap_count(7) => 1
leap_count(8) => 2
```

Reply:
> Could you also add a test in

OK! I added a new test case!
https://github.com/protocolbuffers/protobuf/pull/6676/commits/63837ef911ceeae4d6ca534795d4048258b4494f

Reply:
> Could you also port your code to https://github.com/protocolbuffers/upb/blob/d3762e96cb8910b8281419faee7bd70b323c753d/upb/json/parser.rl#L1707
> and https://github.com/protocolbuffers/protobuf/blob/master/php/ext/google/protobuf/upb.c#L10137

OK! I ported the same code to php!
https://github.com/protocolbuffers/protobuf/pull/6676/commits/3574836addde741d9796fc75ec54eec1808e9359

I'll make another PR for https://github.com/protocolbuffers/upb .